File: internal/model/config_history.go
package model

import "time"

// ConfigHistory 自定义配置的变更历史快照
type ConfigHistory struct {
	ID             uint                     `gorm:"primarykey" json:"id"`
	CustomConfigID uint                     `gorm:"not null;index" json:"custom_config_id"`
	UserID         uint                     `gorm:"not null;index" json:"user_id"`
	Name           string                   `json:"name"`
	ProxyGroups    []map[string]interface{} `gorm:"serializer:json;type:longtext" json:"proxy_groups"`
	Rules          []string                 `gorm:"serializer:json;type:longtext" json:"rules"`
	RuleProviderIDs  []uint                 `gorm:"serializer:json;type:longtext" json:"rule_provider_ids"`
	HostedRuleSetIDs []uint                 `gorm:"serializer:json;type:longtext" json:"hosted_rule_set_ids"`
	CreatedAt      time.Time                `json:"created_at"`
}

File: frontend/src/api/config-history.ts
import client from './client'

export interface ConfigHistory {
  id: number
  custom_config_id: number
  name: string
  proxy_groups: Record<string, unknown>[]
  rules: string[]
  rule_provider_ids: number[]
  hosted_rule_set_ids: number[]
  created_at: string
}

export const configHistoryApi = {
  list: (configId: number) =>
    client
      .get<{ code: number; data: ConfigHistory[] }>(`/custom-configs/${configId}/history`)
      .then((r) => r.data.data),
  restore: (configId: number, historyId: number) =>
    client
      .post<{ code: number; data: unknown }>(`/custom-configs/${configId}/history/${historyId}/restore`)
      .then((r) => r.data.data),
}
